Yubico.YubiKey/src/Yubico/YubiKey/YubiKeyDeviceListener.cs

```diff
@@ -44,83 +45,84 @@ public class YubiKeyDeviceListener : IDisposable
+        private readonly Task _listenTask;
+        private readonly CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private CancellationToken CancellationToken => _tokenSource.Token;
+
+        private bool _isListening;
 
         private YubiKeyDeviceListener()
         {
             _log.LogInformation("Creating YubiKeyDeviceListener instance.");
 
-            _listenerThread = new Thread(ListenForChanges) { IsBackground = true };
-            _isListening = true;
+            SetupDeviceListeners();
 
             _log.LogInformation("Performing initial cache population.");
             Update();
 
-            _listenerThread.Start();
+            _listenTask = ListenForChanges();
         }
 
         internal List<IYubiKeyDevice> GetAll() => _internalCache.Keys.ToList();
 
-        private void ListenForChanges()
-        {
-            using var updateEvent = new ManualResetEvent(false);
-
-            _log.LogInformation("YubiKey device listener thread started. ThreadID is {ThreadID}.", Environment.CurrentManagedThreadId);
-
-            _smartCardListener.Arrived += (s, e) =>
-            {
-                _log.LogInformation("Arrival of smart card {SmartCard} is triggering update.", e.Device);
-                _ = updateEvent.Set();
-            };
-
-            _smartCardListener.Removed += (s, e) =>
-            {
-                _log.LogInformation("Removal of smart card {SmartCard} is triggering update.", e.Device);
-                _ = updateEvent.Set();
-            };
-
-            _hidListener.Arrived += (s, e) =>
-            {
-                _log.LogInformation("Arrival of HID {HidDevice} is triggering update.", e.Device);
-                _ = updateEvent.Set();
-            };
+        private void ListenerHandler(object? sender, EventArgs e) => _semaphore.Release();
 
-            _hidListener.Removed += (s, e) =>
-            {
-                _log.LogInformation("Removal of HID {HidDevice} is triggering update.", e.Device);
-                _ = updateEvent.Set();
-            };
+        private void SetupDeviceListeners()
+        {
+            _smartCardListener.Arrived += ListenerHandler;
+            _smartCardListener.Removed += ListenerHandler;
+            _hidListener.Arrived += ListenerHandler;
+            _hidListener.Removed += ListenerHandler;
+        }
 
+        private async Task ListenForChanges()
+        {
+            _isListening = true;
             while (_isListening)
             {
-                _ = updateEvent.WaitOne();
-                Thread.Sleep(200); // I really dislike sleeps, but here, it does seem like a good idea to give the
-                                   // system some time to quiet down in terms of PnP activity.
-                _ = updateEvent.Reset();
-                Update();
+                try
+                {
+                    await _semaphore.WaitAsync(CancellationToken).ConfigureAwait(false);
```

Comment:
We dropped a log message here. Knowing what started the update process is very useful to know.

Reply:
Added logging.